src/room/planners/init-planner.ts

```diff
@@ -1,18 +1,64 @@
 import {CreepSpawnData} from "../../creeps/creep-spawn-data";
 import {Upgrader} from "../../creeps/roles/upgrader";
+import {CreepBodyBuilder} from "../../creeps/creep-body-builder";
+import {PlannerInterface} from "./planner-interface";
+import {Transport} from "../../creeps/roles/transport";
+import * as _ from "lodash";
 
-export class InitPlanner {
+export class InitPlanner implements PlannerInterface {
     private room: Room;
+    private creepsAssigned = false;
 
     constructor(room: Room) {
         this.room = room;
     }
 
     reassignCreeps() {
-        // TODO reassign creeps to different roles (especially idle ones)
+        if (this.creepsAssigned) {
+            return;
+        }
+        const spawnersNeedingEnergy = this.room.find(FIND_MY_STRUCTURES, {filter: (s:Structure) => {
+                return s['store'] && (s.structureType === STRUCTURE_SPAWN ||
+                    s.structureType === STRUCTURE_EXTENSION) && s['store'].getFreeCapacity(RESOURCE_ENERGY) > 0;
+            }});
+        if (spawnersNeedingEnergy.length > 0) {
+            let transportReassigned = false;
+            _.forEach(this.room.find(FIND_MY_CREEPS), (creep: Creep) => {
+                if (!transportReassigned && creep.memory &&
+                        (!creep.memory['role'] || creep.memory['role'] === Upgrader.KEY)) {
+                    creep.memory['role'] = Transport.KEY;
+                    delete creep.memory['action'];
+                    delete creep.memory['target'];
+                    transportReassigned = true;
+                    this.room.creepCountArray = null;
```

Comment:
It wouldn't be hard to reassign the role in creepCountArray without invalidating the whole thing. Not a big deal, but something to add to that "reassign N from [...] to role" function.

Reply:
Fixed